### PR TITLE
 exchange buffer size per species

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- add trait to get exchange sizes for a species
- add definition for a particle attribute `exchangeMemCfg`
- `Particles.tpp`: use trait `GetExchangeMemCfg` to allow different exchange buffer sizes per species

CC-ing: @HighIander

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
@ax3l Where is the best place to include this feature into our documentation?

File: include/picongpu/particles/traits/GetExchangeMemCfg.hpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        static constexpr uint32_t BYTES_EXCHANGE_Y = picongpu::BYTES_EXCHANGE_Y;
+        static constexpr uint32_t BYTES_EXCHANGE_Z = picongpu::BYTES_EXCHANGE_Z;
+        static constexpr uint32_t BYTES_CORNER = picongpu::BYTES_CORNER;
+        static constexpr uint32_t BYTES_EDGES = picongpu::BYTES_EDGES;
+    };
+
+} // namespace detail
+
+    /** get a memory configuration for species exchange buffer
+     *
```
Comment (psychocoderHPC) [inline include/picongpu/particles/traits/GetExchangeMemCfg.hpp]: documentation missing: `If exchangeMemCfg is not defined than the default memory exchange size in memory.param are used`

File: include/picongpu/simulation_defines/param/memory.param [hunk 2]
```diff
@@ -46,12 +46,22 @@ namespace picongpu
...
-    constexpr uint32_t BYTES_EXCHANGE_Y = 6 * 512 * 1024; //6 MiB
-    constexpr uint32_t BYTES_EXCHANGE_Z = 4 * 256 * 1024; //4 MiB
-    constexpr uint32_t BYTES_CORNER = 8 * 1024; //8 kiB;
-    constexpr uint32_t BYTES_EDGES = 32 * 1024; //32 kiB;
+    /** bytes reserved for species exchange buffer
+     *
+     * This is the default configuration for species exchanges buffer sizes.
+     * The default exchange buffer sizes can be changed per species by adding
+     * the alias exchangeMemCfg with similar members like in DefaultExchangeMemCfg
+     * to it's flag list.
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/memory.param]: `it's` -> its

File: include/picongpu/particles/Particles.tpp [hunk 3]
```diff
@@ -79,44 +80,50 @@ Particles<
     ),
     m_datasetID( datasetID )
 {
-    size_t sizeOfExchanges = 2 * 2 * ( BYTES_EXCHANGE_X + BYTES_EXCHANGE_Y + BYTES_EXCHANGE_Z ) + BYTES_EXCHANGE_X * 2 * 8;
+    using ExchangeMemCfg = GetExchangeMemCfg_t< Particles >;
+
+    size_t sizeOfExchanges = 2 * 2 * (
```
Comment (ax3l) [inline include/picongpu/particles/Particles.tpp]: good time to:
- declare or document what `2*2` and `2*8` stand for
- make `sizeOfExchanges` a `const`

Comment (psychocoderHPC) [reply include/picongpu/particles/Particles.tpp]: I have changed the way how the size of all exchanges is calculated, this should be much better understandable than a comment over the ugly magic numbers.

Comment (psychocoderHPC) [reply include/picongpu/particles/Particles.tpp]: btw: Maybe the calculated amount was wrong for 3D or 2D because the old equation was equal for 2D and 3D

**Pull request merged**